src/core/primitives.rs

```diff
@@ -451,3 +451,52 @@ fn refresh_coordinate_keys(
         })
         .collect::<RevisionVec<_, _>>()
 }
+
+/// Attempts opening the Covercrypt encapsulation using the given USK. Returns
+/// the encapsulated key and associated rights upon success, otherwise returns `None`.
+pub fn full_decaps(
+    usk: &UserSecretKey,
+    encapsulation: &XEnc,
+) -> Result<Option<Vec<(Right, Secret<SHARED_SECRET_LENGTH>)>>, Error> {
+    // A = ⊙ _i (α_i. c_i)
+    let A = usk
+        .id
+        .iter()
+        .zip(encapsulation.c.iter())
+        .map(|(marker, trap)| trap * marker)
+        .fold(EcPoint::identity(), |mut acc, elt| {
+            acc = &acc + &elt;
+            acc
+        });
+
+    for enc in &encapsulation.encapsulations {
+        let mut rights_list: Vec<(Right, Secret<SHARED_SECRET_LENGTH>)> = Vec::new();
+        for secret in usk.secrets.flat_iter() {
+            let S = match (secret.1, enc) {
+                (RightSecretKey::Hybridized { sk, dk }, Encapsulation::Hybridized { E, F }) => {
+                    let mut K1 = h_hash(R25519::session_key(&sk, &A)?);
+                    let K2 = MlKem512::dec(&dk, E)?;
+                    let S = xor_3(F, &K1, &K2);
+                    K1.zeroize();
+                    S
+                }
+                (RightSecretKey::Classic { sk }, Encapsulation::Classic { F }) => {
+                    let K1 = h_hash(R25519::session_key(&sk, &A)?);
+                    xor_2(F, &K1)
+                }
+                (RightSecretKey::Hybridized { .. }, Encapsulation::Classic { .. })
+                | (RightSecretKey::Classic { .. }, Encapsulation::Hybridized { .. }) => {
+                    continue;
+                }
+            };
+
+            let (tag, ss) = j_hash(&S, &encapsulation.c, &encapsulation.encapsulations)?;
+
+            if tag == encapsulation.tag {
+                rights_list.push((secret.0.clone(), ss));
+            }
```

Comment:
You should jump to the next right as soon as you find a key in this right's history that is able to open an encapsulation.